### PR TITLE
[WIP] TP + FSDP Integration change

### PR DESCRIPTION
**Pull request by fduwjj**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82581

This is PR or diff is not aiming for land now.
Context: We want to have FSDP + TP 2-D Parallel enabled for transformer based model.
Original PR: https://github.com/pytorch/pytorch/pull/82166
Andrew's Rebased commit: https://github.com/awgu/pytorch/commit/5b7ef1a6ea2e802fb6f1276baea6a030090320f4

This PR includes the use of Distributed Tensor. Down the road, we can discuss whether we only want to enable it for distributed tensor or not. For now, I just keep the logic both for sharded tensor and distributed tensor.

Will update a unit test for TP + FSDP using DT soon.

Differential Revision: [D38266184](https://our.internmc.facebook.com/intern/diff/D38266184/)

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -14,9 +14,15 @@
 )
 
 import torch
+import torch.distributed as dist
+import torch.distributed._shard.sharding_spec as shard_spec
 import torch.nn as nn
 import torch.nn.functional as F
 from torch import Tensor
+from torch.distributed._shard.sharded_tensor.api import ShardedTensor
+from spmd.tensor import Tensor as DistributedTensor, DeviceMesh
```
Comment (kumpera) [inline torch/distributed/fsdp/flat_param.py:23]: this needs to guarded in a try/except so people can use it without pipy installed.

Comment (fduwjj) [reply torch/distributed/fsdp/flat_param.py:23]: sure.

Comment (fduwjj) [reply torch/distributed/fsdp/flat_param.py]: Added.

**Comment by fduwjj** [comment]
Not needed.

**Pull request closed without merging**